File: pkg/services/exploremap/crdt/orset.go
package crdt

import (
	"encoding/json"
)

// ORSet represents an Observed-Remove Set CRDT
type ORSet struct {
	adds    map[string]map[string]bool // element -> set of tags
	removes map[string]bool            // set of removed tags
}

// ORSetJSON is the JSON representation of ORSet
type ORSetJSON struct {
	Adds    map[string][]string `json:"adds"`
	Removes []string            `json:"removes"`
}

// NewORSet creates a new empty OR-Set
func NewORSet() *ORSet {
	return &ORSet{
		adds:    make(map[string]map[string]bool),
		removes: make(map[string]bool),
	}
}

// Add adds an element to the set with a unique tag
func (s *ORSet) Add(element, tag string) {
	if s.adds[element] == nil {
		s.adds[element] = make(map[string]bool)
	}
	s.adds[element][tag] = true
}

// Remove removes an element from the set
// Only removes the specific tags that were observed
func (s *ORSet) Remove(element string, observedTags []string) {
	for _, tag := range observedTags {
		s.removes[tag] = true
	}

	// Clean up the element's tags
	if elementTags, exists := s.adds[element]; exists {
		for _, tag := range observedTags {
			delete(elementTags, tag)
		}

		// If no tags remain, remove the element entry
		if len(elementTags) == 0 {
			delete(s.adds, element)
		}
	}
}

// Contains checks if an element is in the set
// Element is present if it has at least one non-removed tag
func (s *ORSet) Contains(element string) bool {
	tags, exists := s.adds[element]
	if !exists || len(tags) == 0 {
		return false
	}

	// Element is present if it has at least one tag that hasn't been removed
	for tag := range tags {
		if !s.removes[tag] {
			return true
		}
	}
	return false
}

// GetTags returns all tags for an element (including removed ones)
func (s *ORSet) GetTags(element string) []string {
	tags, exists := s.adds[element]
	if !exists {
		return []string{}
	}

	result := make([]string, 0, len(tags))
	for tag := range tags {
		result = append(result, tag)
	}
	return result
}

// Values returns all elements currently in the set
func (s *ORSet) Values() []string {
	result := make([]string, 0, len(s.adds))

	for element, tags := range s.adds {
		// Include element if it has at least one non-removed tag
		for tag := range tags {
			if !s.removes[tag] {
				result = append(result, element)
				break
			}
		}
	}

	return result
}

// Size returns the number of elements in the set
func (s *ORSet) Size() int {
	return len(s.Values())
}

// IsEmpty checks if the set is empty
func (s *ORSet) IsEmpty() bool {
	return s.Size() == 0
}

// Merge merges another OR-Set into this one
// Takes the union of all adds and removes
func (s *ORSet) Merge(other *ORSet) {
	// Merge adds (union of all tags)
	for element, otherTags := range other.adds {
		if s.adds[element] == nil {
			s.adds[element] = make(map[string]bool)
		}
		for tag := range otherTags {
			s.adds[element][tag] = true
		}
	}

	// Merge removes (union of all removed tags)
	for tag := range other.removes {
		s.removes[tag] = true
	}

	// Clean up elements with all tags removed
	for element, tags := range s.adds {
		hasLiveTag := false
		for tag := range tags {
			if !s.removes[tag] {
				hasLiveTag = true
				break
			}
		}
		if !hasLiveTag {
			delete(s.adds, element)
		}
	}
}

// Clone creates a deep copy of the OR-Set
func (s *ORSet) Clone() *ORSet {
	clone := NewORSet()

	// Deep copy adds
	for element, tags := range s.adds {
		clone.adds[element] = make(map[string]bool)
		for tag := range tags {
			clone.adds[element][tag] = true
		}
	}

	// Deep copy removes
	for tag := range s.removes {
		clone.removes[tag] = true
	}

	return clone
}

// Clear removes all elements from the set
func (s *ORSet) Clear() {
	s.adds = make(map[string]map[string]bool)
	s.removes = make(map[string]bool)
}

// MarshalJSON implements json.Marshaler
func (s *ORSet) MarshalJSON() ([]byte, error) {
	adds := make(map[string][]string)
	for element, tags := range s.adds {
		tagList := make([]string, 0, len(tags))
		for tag := range tags {
			tagList = append(tagList, tag)
		}
		adds[element] = tagList
	}

	removes := make([]string, 0, len(s.removes))
	for tag := range s.removes {
		removes = append(removes, tag)
	}

	return json.Marshal(ORSetJSON{
		Adds:    adds,
		Removes: removes,
	})
}

// UnmarshalJSON implements json.Unmarshaler
func (s *ORSet) UnmarshalJSON(data []byte) error {
	var tmp ORSetJSON
	if err := json.Unmarshal(data, &tmp); err != nil {
		return err
	}

	s.adds = make(map[string]map[string]bool)
	for element, tags := range tmp.Adds {
		s.adds[element] = make(map[string]bool)
		for _, tag := range tags {
			s.adds[element][tag] = true
		}
	}

	s.removes = make(map[string]bool)
	for _, tag := range tmp.Removes {
		s.removes[tag] = true
	}

	return nil
}

File: public/app/features/explore-map/crdt/orset.ts
/**
 * Observed-Remove Set (OR-Set) CRDT implementation
 *
 * A set that handles concurrent add/remove operations correctly.
 * Each element is tagged with unique identifiers, and removes only
 * affect the specific tags they observed.
 *
 * Properties:
 * - Add-wins semantics: concurrent add/remove results in element being present
 * - Idempotent operations: applying same operation multiple times is safe
 * - Commutative: operations can be applied in any order
 *
 * Used for tracking which panels exist on the canvas.
 */

export interface ORSetJSON<T> {
  adds: Record<string, string[]>;  // element -> array of unique tags
  removes: string[];                // array of removed tags
}

export class ORSet<T extends string = string> {
  private adds: Map<T, Set<string>>;  // element -> set of unique tags
  private removes: Set<string>;        // set of removed tags

  constructor() {
    this.adds = new Map();
    this.removes = new Set();
  }

  /**
   * Add an element to the set with a unique tag
   * The tag should be globally unique (e.g., operation ID)
   *
   * @param element - The element to add
   * @param tag - Unique identifier for this add operation
   */
  add(element: T, tag: string): void {
    if (!this.adds.has(element)) {
      this.adds.set(element, new Set());
    }
    this.adds.get(element)!.add(tag);
  }

  /**
   * Remove an element from the set
   * Only removes the specific tags that were observed
   *
   * @param element - The element to remove
   * @param observedTags - The tags that were observed when the remove was issued
   */
  remove(element: T, observedTags: string[]): void {
    for (const tag of observedTags) {
      this.removes.add(tag);
    }

    // Clean up the element's tags
    const elementTags = this.adds.get(element);
    if (elementTags) {
      for (const tag of observedTags) {
        elementTags.delete(tag);
      }

      // If no tags remain, remove the element entry
      if (elementTags.size === 0) {
        this.adds.delete(element);
      }
    }
  }

  /**
   * Check if an element is in the set
   * Element is present if it has at least one non-removed tag
   *
   * @param element - The element to check
   * @returns true if element is in the set
   */
  contains(element: T): boolean {
    const tags = this.adds.get(element);
    if (!tags || tags.size === 0) {
      return false;
    }

    // Element is present if it has at least one tag that hasn't been removed
    for (const tag of tags) {
      if (!this.removes.has(tag)) {
        return true;
      }
    }
    return false;
  }

  /**
   * Get all tags for an element (including removed ones)
   *
   * @param element - The element to get tags for
   * @returns Array of tags, or empty array if element not found
   */
  getTags(element: T): string[] {
    const tags = this.adds.get(element);
    return tags ? Array.from(tags) : [];
  }

  /**
   * Get all elements currently in the set
   *
   * @returns Array of elements
   */
  values(): T[] {
    const result: T[] = [];
    for (const [element, tags] of this.adds.entries()) {
      // Include element if it has at least one non-removed tag
      for (const tag of tags) {
        if (!this.removes.has(tag)) {
          result.push(element);
          break;
        }
      }
    }
    return result;
  }

  /**
   * Get the number of elements in the set
   */
  size(): number {
    return this.values().length;
  }

  /**
   * Check if the set is empty
   */
  isEmpty(): boolean {
    return this.size() === 0;
  }

  /**
   * Merge another OR-Set into this one
   * Takes the union of all adds and removes
   *
   * @param other - The OR-Set to merge
   * @returns This OR-Set (for chaining)
   */
  merge(other: ORSet<T>): this {
    // Merge adds (union of all tags)
    for (const [element, otherTags] of other.adds.entries()) {
      if (!this.adds.has(element)) {
        this.adds.set(element, new Set());
      }
      const myTags = this.adds.get(element)!;
      for (const tag of otherTags) {
        myTags.add(tag);
      }
    }

    // Merge removes (union of all removed tags)
    for (const tag of other.removes) {
      this.removes.add(tag);
    }

    // Clean up elements with all tags removed
    for (const [element, tags] of this.adds.entries()) {
      let hasLiveTag = false;
      for (const tag of tags) {
        if (!this.removes.has(tag)) {
          hasLiveTag = true;
          break;
        }
      }
      if (!hasLiveTag) {
        this.adds.delete(element);
      }
    }

    return this;
  }

  /**
   * Create a copy of this OR-Set
   */
  clone(): ORSet<T> {
    const copy = new ORSet<T>();

    // Deep copy adds
    for (const [element, tags] of this.adds.entries()) {
      copy.adds.set(element, new Set(tags));
    }

    // Deep copy removes
    copy.removes = new Set(this.removes);

    return copy;
  }

  /**
   * Clear all elements from the set (for testing/reset)
   */
  clear(): void {
    this.adds.clear();
    this.removes.clear();
  }

  /**
   * Serialize to JSON for network transmission or storage
   */
  toJSON(): ORSetJSON<T> {
    const adds: Record<string, string[]> = {};
    for (const [element, tags] of this.adds.entries()) {
      adds[element] = Array.from(tags);
    }

    return {
      adds,
      removes: Array.from(this.removes),
    };
  }

  /**
   * Deserialize from JSON
   */
  static fromJSON<T extends string = string>(json: ORSetJSON<T>): ORSet<T> {
    const set = new ORSet<T>();

    // Restore adds
    for (const [element, tags] of Object.entries(json.adds)) {
      set.adds.set(element as T, new Set(tags));
    }

    // Restore removes
    set.removes = new Set(json.removes);

    return set;
  }

  /**
   * Get debug information about the set
   */
  debug(): {
    elements: T[];
    totalTags: number;
    removedTags: number;
    rawAdds: Map<T, Set<string>>;
    rawRemoves: Set<string>;
  } {
    let totalTags = 0;
    for (const tags of this.adds.values()) {
      totalTags += tags.size;
    }

    return {
      elements: this.values(),
      totalTags,
      removedTags: this.removes.size,
      rawAdds: this.adds,
      rawRemoves: this.removes,
    };
  }
}
